tools/java/org/unicode/cldr/api/PathMatcher.java

```diff
@@ -0,0 +1,276 @@
+// © 2019 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+// © 2019 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+// © 2019 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+package org.unicode.cldr.api;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkPositionIndex;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static org.unicode.cldr.api.AttributeKey.keyOf;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * An immutable matcher for {@link CldrPath} instances. A path matcher specification looks like
+ * {@code "//ldml/foo/*[@x="z"]/bar[@y=*]"}, where element names and attribute values can be
+ * wildcards. No capturing of wildcard values is done, and currently wildcards are limited to the
+ * entire field (element name, or attribute value).
+ *
+ * Examples of use:
+ *
+ * <h2>Matching an exact path</h2>
+ * <pre>{@code
+ * private static final PathMatcher ALIAS_TO_XX =
+ *     PathMatcher.of("//supplementalData/likelySubtags/likelySubtag[@from=*][@to="xx"]");
+ *
+ * if (ALIAS_TO_XX.matches(value.getPath())) {
+ *     // Value is a subtag alias to locale ID "xx".
+ * }
+ * }</pre>
+ *
+ * <h2>Matching path prefixes and hierarchical matching</h2>
+ * <pre>{@code
+ * private static final PathMatcher KEY = PathMatcher.of("//ldmlBCP47/keyword/key[@name=*]");
+ * private static final PathMatcher TYPE = key.extendBy("type[@name=*]");
+ *
+ * if (KEY.matchesPrefixOf(path)) {
+ *     if (TYPE.locallyMatches(path)) {
+ *         // Value is a BCP47 type.
+ *     }
+ * }
+ * }</pre>
+ */
+public final class PathMatcher {
+    private static final Pattern ROOT_PATH_SPEC = Pattern.compile("//(\\w+)(/|$)");
+
+    /**
+     * Parses the full or partial (prefix) path specification into a matcher. Path specifications
+     * given to this method must start with {@code "//xxx/..."} where {@code xxx} is a known CLDR
+     * XML path prefix (such as {@code "ldml").
+     */
+    public static PathMatcher of(String pathSpec) {
```

Comment:
I assume that you are moving this code from ICU, which means it's been reviewed. I am not looking super thoroughly.

Reply:
Mostly. It's subject to the API changes in the doc we discussed. I hadn't realized when I first sent this out that Mark wanted a "proper" review (and I won't now submit until that's sorted).